### PR TITLE
Generalize partial sum product to Markov models

### PR DESCRIPTION
**Pull request by ordabayevy**
This attempts to generalize `partial_sum_product` to support elimination of markov variables following discussion with @eb8680 and @fritzo at pyro-ppl/pyro#2687.

**TODO**
- [x] `modified_partial_sum_product` that eliminates first-order markov variables using parallel-scan algorithm.
- [x] Add more tests
- [ ] <s>Rename `modified_partial_sum_product` or replace `partial_sum_product` by it?</s>

**Tests**
- [x] `_expected_modified_partial_sum_product` unrolls one markov  dim at a time
- [x] `test_modified_partial_sum_product_0`
- [x] `test_modified_partial_sum_product_1`
- [x] `test_modified_partial_sum_product_2`
- [x] `test_modified_partial_sum_product_3`
- [x] `test_modified_partial_sum_product_4`
- [x] `test_modified_partial_sum_product_5`
- [x] `test_modified_partial_sum_product_6`
- [x] `test_modified_partial_sum_product_7`: xfail `x` and `y` markov vars with different ordinals
- [x] `test_modified_partial_sum_product_8`
- [x] `test_modified_partial_sum_product_9`
- [x] `test_modified_partial_sum_product_10`
- [x] `test_modified_partial_sum_product_11`
- [x] `test_modified_partial_sum_product_12`: xfail `w_curr` in time plate but lower ordinal than `y` and `x` markov vars
- [x] `test_modified_partial_sum_product_13`
- [x] `test_modified_partial_sum_product_14`
- [x] `test_modified_partial_sum_product_15`: xfail `y` depends on two time dims `time` and `tones`
- [x] `test_modified_partial_sum_product_16`: `x_prev->y_curr` and `y_prev->x_curr`

**Note**
- Markov sites cannot be within local plates inside (see tests 2 and 10)

**Update**
- Added unrolling of markov vars in local plates to handle tests 2 and 10

**Useful links**
- pyro-ppl/funsor#167
- pyro-ppl/funsor#204
- pyro-ppl/funsor#293 
- pyro-ppl/numpyro#686
- pyro-ppl/funsor#399

### CODE REVIEW

**Comment by ordabayevy** [comment]
Not sure why `test/test_sum_product.py::test_markov_partial_sum_product_multi` is failing. The test passes locally.

**Comment by eb8680** [comment]
@ordabayevy awesome progress!  I'm impressed, especially considering how poorly documented the code in `funsor.sum_product` is 😆 

I think in this first PR we should focus on getting a basic version of the algorithm correct, ignoring the optimizer and window size > 1 as well as any interface changes to `sarkka_bilmes_product`.  Once you have that and we have collected a large number of test cases, we should be able to work our way up to the full version fairly quickly via a sequence of smaller PRs with simpler conceptual transitions between steps. 

To that end, can I suggest something like the following sequence of PRs?

1. A version of `modified_partial_sum_product` that works for time lag 1, and tests that compare it to `partial_sum_product` applied to the unrolled factors.  I think you've mostly got that here, but let's remove `modified_sequential_sum_product` from this first PR in favor of `sequential_sum_product` and focus on adding lots of test cases.
2. Making `modified_partial_sum_product` use `MarkovProduct` and additional tests that verify that it works with `funsor.optimize.apply_optimizer`
3. A version of `modified_partial_sum_product` that uses `sarkka_bilmes_product` directly, and tests that compare it to `partial_sum_product` applied to unrolled factors
4. Tests verifying that the `sarkka_bilmes_product`-based `modified_partial_sum_product` works with `funsor.optimize.apply_optimizer`, and any fixes necessary to make them pass (e.g. a first-class `funsor.Funsor` wrapper for `sarkka_bilmes_product` analogous to `MarkovProduct`)
5. A nice Funsor example or two to show off the new algorithm
6. Optional: Any interface changes to `partial_sum_product` or `sarkka_bilmes_product` (e.g. your `modified_sequential_sum_product` interface)
7. Optional: Additional enhancements/features (e.g. support for partial windows)
8. Optional: removal of less general algorithms implemented in previous steps
9. Integration with Pyro for use with enumeration

**Comment by eb8680** [comment]
> Not sure why `test/test_sum_product.py::test_markov_partial_sum_product_multi` is failing. The test passes locally.

It looks like a floating point precision error caused by unrolling, which can happen occasionally since you're generating test data (i.e. factors) randomly via `funsor.testing.random_tensor`.  Reducing the size of the test data time axis might help, or you can tweak the test tolerance by changing the `atol` and `rtol` arguments to `assert_close` in your tests.

**Comment by ordabayevy** [comment]
@eb8680 thanks for your feedback! Keeping it simple for this PR is a good idea. I have removed `modified_sequential_sum_product` to a separate PR #399 in case it might be useful in the future.

**Comment by ordabayevy** [comment]
Hi @eb8680 . I have added tests `test_partial_sum_product_hmm_example_[1-4]` based on models given in HMM examples and `test_partial_sum_product_hmm_example_[8-10]` are their variations.  Expected values are calculated by `_expected_hmm_example` via unrolling time slices as you mentioned before. Current algorithm doesn't work for tests `_2` and `_10` because markov sites are inside local plates. Seems like these local plates have to be unrolled  before using `sequential_sum_product`(?). I'll try to add unrolling step to see if it works.

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:174]: Note this `y_curr` and the `sequences` plate means this test does not correspond to model 0, although I think it's still a reasonable test case since factors `x` and `y` are completely disconnected.

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:223]: When are the plates `tones` and `sequences` being eliminated by `modified_partial_sum_product`? It should be before and after `time` respectively, rather than both before or both after.

File: test/test_sum_product.py
Comment (ordabayevy) [reply test/test_sum_product.py:274]: Yeah, I kept `y_prev` and `y_curr` and unobserved in order to consider more general cases. As I mentioned above we might be able to generalize to these cases by using `sequential_sum`.

File: test/test_sum_product.py
Comment (ordabayevy) [reply test/test_sum_product.py:223]: For different combinations of `vars1` and `vars2` it follows that order to eliminate plates. Second case is when `tones` removed before `time` and third case is `sequences` eliminated after `time`.

File: funsor/sum_product.py
Comment (ordabayevy) [inline funsor/sum_product.py:164]: @eb8680 @fritzo thanks for helping with the code! Here I made some changes to its structure. `sequential_sum_product` is in a separate logic block (and that's the only extra code compared to `partial_sum_product`):
- since `sequential_sum_product` is called at most only once I feel like there is no need for yet another function `_markov_product`
- now it is easier to test for intractable cases (line 157)
- additionally we can easily check that there is at most one markov plate per branch (line 154)

WDYT?

Comment (eb8680) [reply funsor/sum_product.py:164]: > Here I made some changes to its structure. sequential_sum_product is in a separate logic block (and that's the only extra code compared to partial_sum_product)

Makes sense to me.  I do wonder if there's a simpler way to formulate the intractability conditions that unifies them into a single condition, but I guess we don't need to figure that out now.

File: funsor/sum_product.py
Comment (eb8680) [inline funsor/sum_product.py:159]: Can you explain this TODO?

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:174]: Thanks for renaming these

File: funsor/sum_product.py
Comment (eb8680) [inline funsor/sum_product.py:100]: Make the docstring a bit more informative, e.g. "Generalization of the tensor variable elimination algorithm of `funsor.sum_product.partial_sum_product` to handle markov dimensions in addition to plate dimensions.  Markov dimensions are eliminated efficiently using the parallel-scan algorithm in `funsor.sum_product.sequential_sum_product`."

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:113]: Have you verified manually that the unrolling logic in this function produces the right factors for a few of the test cases? It looks correct to me but it's still somewhat complicated and all of the new tests depend on it.

**Review by eb8680 — COMMENTED** [review]
The algorithm and tests mostly look good, except for a few nits. Can you also add a couple additional test cases?

1. Like `test_modified_partial_sum_product_5`, but with a global sum variable coupling the two time dimensions
2. Something exercising the `len(time) == 1` intractability case

File: funsor/sum_product.py
Comment (ordabayevy) [reply funsor/sum_product.py]: This is irrelevant to this PR. I removed it.
(I think it would be simpler to pass just the name of the time dim to the `sequential_sum_product` instead of `time_var`)

File: test/test_sum_product.py
Comment (ordabayevy) [reply test/test_sum_product.py]: I reworked the unrolling logic (it unrolls one markov dim at a time for all factors now). [Here](https://docs.google.com/presentation/d/1K2aG9reNF0KWehGzxI8b7SsMqW-FaOPY1tfdb2r2vnA/edit?usp=sharing) I have manually unrolled markov dims for tests 11 and 14 and compared to the unrolled factors by this function. It is complicated and would be nice if someone else also verified it.

**Review by ordabayevy — COMMENTED** [review]
Sorry for the late reply. New added tests:

1. `test_modified_partial_sum_product_13` with a global sum variable coupling the two time dimensions
2. This one is tricky. I came up with `test_modified_partial_sum_product_14` but then realized that it actually might be tractable. With some changes in the code of the algorithm it seems to pass this test. I also have checked its unrolling manually [here]( actually).

The added docstring is as you typed it.

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:1062]: I guess the reason this doesn't fail as intractable is that you haven't included `y` in the `time` plate, so it's treating `tones` as nested entirely within a single step of `time`.

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:965]: `test_modified_partial_sum_product_13` looks good, thanks.

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:115]: Can you explain the function of this new `local_markov_var_dict` argument? It doesn't seem to be used in any of the tests except `test_modified_partial_sum_product_14`.

File: test/test_sum_product.py
Comment (eb8680) [reply test/test_sum_product.py:1062]: For this to fail as intractable, you might have to include two markov variables `y_prev_tones` and `y_prev_time` for `y` in `h`, one for each time variable, and put them in `plate_to_step` appropriately:
```py
plate_to_step = dict({
    "sequences": {},
    "time": {"x_prev": "x_curr", "y_time_prev": "y_curr"},
    "tones": {"y_prev": "y_curr", "y_tones_prev": "y_curr"}
})
```

File: test/test_sum_product.py
Comment (eb8680) [inline test/test_sum_product.py:176]: To clarify, is the purpose of these two new substitutions to handle cases like the current, tractable version of `test_modified_partial_sum_product_14` where one time variable is completely nested within a single step of another?

**Comment by eb8680** [comment]
OK, I think I understand the unrolling logic now modulo a couple of questions, and it looks correct to me along with your hand-unrolled examples, at least for the test cases in this PR.  

Perhaps in a future PR It would be a good idea to provide a general `funsor.sum_product.unroll` implementation that converts a `modified_partial_sum_product` problem instance (i.e. the arguments to `modified_partial_sum_product`) into a sum-product problem instance with no time variables or plates.

I'm willing to merge this PR once we have an xfailing test covering `len(time) > 1`.

File: test/test_sum_product.py
Comment (ordabayevy) [reply test/test_sum_product.py:176]: Yep, that's right.

File: funsor/sum_product.py
Comment (ordabayevy) [inline funsor/sum_product.py:131]: Collects all time dims that a markov var depends on in a set. In line 162 raise intractable error if the set size is not equal 1.

File: funsor/sum_product.py
Comment (ordabayevy) [inline funsor/sum_product.py:153]: `| markov_prod_vars` is added to `_partition` for `test_modified_partial_sum_product_16` to pass. Otherwise leaf factors (f2 and f3) are partitioned into two different groups.

File: test/test_sum_product.py
Comment (ordabayevy) [reply test/test_sum_product.py:1062]: See `test_modified_partial_sum_product_15`

**Review by ordabayevy — COMMENTED** [review]
Two more tests:

1. test 15 intractable because `y` depends on two time dims `time` and `tones`
2. test 16 where `y_prev->x_curr` and `x_prev->y_curr`

**Comment by eb8680** [comment]
Last 2 tests look good. One final task before merging: can you run `test_sum_product.py` locally with all of your new tests under the JAX backend (`FUNSOR_BACKEND=jax pytest test_sum_product.py ...`)? 

We don't include those on Travis because they're so slow and memory-intensive, so you'll need to run them locally to catch any potential bugs.

**Comment by ordabayevy** [comment]
@eb8680 passed `FUNSOR_BACKEND=jax pytest test_sum_product.py -n auto`:

<details>
```
............................................................................................. [  3%]
............................................................................................. [  7%]
............................................................................................. [ 11%]
............................................................................................. [ 14%]
............................................................................................. [ 18%]
............................................................................................. [ 22%]
............................................................................................. [ 25%]
............................................................................................. [ 29%]
............................................................................................. [ 33%]
.............................................................................................. [ 37%]
............................................................................................. [ 40%]
............................................................................................. [ 44%]
............................................................................................. [ 48%]
............................................................................................. [ 51%]
............................................................................................. [ 55%]
............................................................................................. [ 59%]
............................................................................................. [ 62%]
............................................................................................. [ 66%]
............................x...x....x.x..x....x.............x.x..x.......................... [ 70%]
...........................................................x...................x....x..x..x.. [ 74%]
.x.xxxx.......xx.x.....xxxxx.xxxxxx.................x...xx..x......x...x...x..x...........xx. [ 77%]
.......x..............x...x.....x.x.....x.......x.....x...x.xxxx.....x.x.x.x...x.x.......xxxx [ 81%]
xxx..x.xxx...x.xxx...xx.x...x..x.x.x.x...x.x.xx.........x..x......................x......x... [ 85%]
......x.....x...xx.x.........x..x...x.x.x..x...xx.xxxx.x.x..xx.xx..xx.xx..................x.. [ 88%]
.x..........x..........x...x..x..x...xx...xx.......x...x.......xx..x...xxx..x..x.xxxx.xxxxxxx [ 92%]
xxx..xx.xxx.......x.xxx..xx..x..x..xxxxxx..x......x..x.....x......x....x..........x..x....... [ 96%]
............x.....x............x..........................x...................x.........x.... [ 99%]
.                                                                                             [100%]
=========================== 2325 passed, 188 xfailed in 258.88s (0:04:18) ===========================
```
</details>

**Review by eb8680 — APPROVED** [review]
Looks good to me.  Nice work!

**Comment by ordabayevy** [comment]
Yay! Thanks for all the help!

**Pull request merged**
